### ISSUE TITLE
[System.Text.Json] Null ref when attempting to deserialize an json object as an array

### ISSUE THREAD

**Issue by pranavkm** [dotnet/corefx#41839]
Ran in to this by accident. Not entirely sure if it's already fixed \ tracked since there's a few that are pretty similar to this one:

```C#
using System.Text.Json;

namespace ConsoleApp3
{
    class Program
    {
        static void Main(string[] args)
        {
            var json = JsonSerializer.Deserialize<int[]>("{ \"a\": \"b\" }");
        }
    }
}
```

```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Text.Json.JsonClassInfo.GetProperty(ReadOnlySpan`1 propertyName, ReadStackFrame& frame)
   at System.Text.Json.JsonSerializer.HandlePropertyName(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& readStack)
   at System.Text.Json.JsonSerializer.ReadCore(Type returnType, JsonSerializerOptions options, Utf8JsonReader& reader)
   at System.Text.Json.JsonSerializer.ParseCore(String json, Type returnType, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
```

**Comment by layomia** [comment]
Was able to repro this. @steveharter, @ahsonkhan looks like 3.1 bug?

**Comment by ahsonkhan** [comment]
> looks like 3.1 bug?

Yes.

It looks like there's a class of null refs that happen whenever the T we are deserializing to doesn't match the input payload. We should look at the previous fixes for similar cases and make sure we are guarding against the whole set.

**Comment by ahsonkhan** [comment]
Re-opening until its fixed for 3.1 or we update the milestone.
